[5.7] making addFailure public

Eugene-Melbourne opened the pull request:
Based on [this stackoverflow question](https://stackoverflow.com/questions/51663977/laravel-custom-validation-failure-with-a-nice-message), this [issue](https://github.com/laravel/framework/issues/25068) and improved in accordance with [this pr improvement request](https://github.com/laravel/framework/pull/25058)

We need a way to GENERATE error messages based on  `\resources\lang\en\validation.php` configuration.

This functionality is sitting in addFailure function. But this function is protected.

So instead of doing:

     $validator->addFailure('my_number', 'not_in', []);

Laravel users may to do:

    $validator->errors()->add('my_number', __('validation.not_in', ['attribute' => __('validation.attributes.my_number')]));

1) But it is not the same because it will ignore `custom` section of validation.php.
2) Some rules has more parameters, not only `attribute`.

So it is quite complex to reproduce `addFailure` functionality using `errors()->add(...` mechanics.

My pull request is  merely making addFailure public.

Eugene-Melbourne committed:
Making addFailure function - public and setting up defaults for the last parameter.

This gives users an option to use existing functionality to generate and add new error message. The existing functionality takes care of translation and ":attribute" naming. It also simplifies the use of \resources\lang\en\validation.php file as a single source of truth for an error messages.

Eugene-Melbourne committed:
test cases around Making addFailure function

creating test cases

Eugene-Melbourne committed:
formatting

changed formatting to satisfy styleci.io

Eugene-Melbourne committed:
comment changes

GrahamCampbell commented:
Thanks, but please send to the 5.7 branch.

tillkruss commented:
Changed the base.

tillkruss commented on src/Illuminate/Validation/Validator.php:
```diff
@@ -584,8 +584,12 @@ protected function shouldStopValidating($attribute)
     {
+        if (! $this->messages) {
+            $this->passes();
+        }
```
I don't see any tests for this logic.

tillkruss commented on src/Illuminate/Validation/Validator.php:
```diff
@@ -584,8 +584,12 @@ protected function shouldStopValidating($attribute)
+        if (! $this->messages) {
+            $this->passes();
+        }
+
```
I don't see any tests for this logic. What is this doing?

tillkruss replied:
We need a test for that logic to make sure it works either way.

Eugene-Melbourne replied:
@tillkruss. Thank you.  I agree the tests are needed. My tests are asserting it (see section 2 of my answer). Please, specify what should I do?

Eugene-Melbourne replied:
@tillkruss , I just checked one more time. My test cases are covering this functionality. If the code that you mention is deleted the test case 'testAddFailureIsFunctional'  is getting an error. `Error: Call to a member function add() on null`.

Eugene-Melbourne replied:
I looked through other test cases, there are plenty of tests around `$validator->passes();`.

devcircus commented:
There's much more going on here than just making a method public. Visibility, signature, and logic is all changing.

Eugene-Melbourne committed:
removed class reflection

removed all of those tests that are reflecting into the class to make sure it's public 😬

Eugene-Melbourne commented:
@devcircus , yes there is not just `private` to `public` change. There are other changes involved to be sure that public method may be called from outside world safely. 

A) There is an initialisation for messages property added, because this is what all other similar public methods have. For example see functions:

1) public function messages()
2) public function invalid()
3) public function valid()

B) A function parameter named $parameters can be an empty array, it is often an empty array. It is a common design pattern in Laravel to set a default parameter in such cases.
For example, the most thought through class of Laravel:

1)  file `src\Illuminate\Contracts\Container\Container.php` has a declaration

    public function make($abstract, array $parameters = []);

2) the same class 'src\Illuminate\Validation\Validator.php' has this approach in declaring a constructor.

Eugene-Melbourne commented:
@taylorotwell. I have done what you requested. It is an honour for me.  Thank you.

GrahamCampbell commented on tests/Validation/ValidationAddFailureTest.php:
```diff
@@ -0,0 +1,41 @@
+     *
+     * @return Validator
+     */
+    public function makeValidator(): Validator
```
Laravel does not use return types in the code like this.

Eugene-Melbourne replied:
Laravel 5.6 Server Requirements [PHP >= 7.1.3](https://laravel.com/docs/5.6), meaning it is legal, 
 you can even use nullable types there if needed. Plus, it is a test case. It is the best place for introducing innovations.

Eugene-Melbourne replied:
Are you sure you want me to remove the returning type declaration?

36864 replied:
No other method in the framework uses return type hints. It would be weird to have a single method using them.

Pretty much everything has docblock declarations, so it's not like we don't know what's being returned.

GrahamCampbell replied:
Laravel's code style conventions forbid using php 7 style return types.

Eugene-Melbourne replied:
Done, return type declaration removed. P.S. It is sad that no method in the framework uses return type hints. :(

Eugene-Melbourne replied:
@GrahamCampbell ,  Thank you for explanation. The requested change is done.

Eugene-Melbourne replied:
@GrahamCampbell , I still can see a big red button with a cross from your change request. Does it mean you have not accept my change? I really don't know.

tillkruss commented on src/Illuminate/Validation/Validator.php:
```diff
@@ -584,8 +584,12 @@ protected function shouldStopValidating($attribute)
+    public function addFailure($attribute, $rule, $parameters = [])
     {
+        if (! $this->messages) {
+            $this->passes();
```
We still need a test for this line.

Eugene-Melbourne replied:
Please help me. Give me a hint on what is expected. My test is asserting that this line is there and it does what it does. What else can I do?

If it was not a `new MessageBag` in the `function passes()`, but instead it was `app()->make(MessageBag...` I could mock it and validate that the passes() function was run. But in this situation I don't see what can be done. Please advise me what to do. Thank you.

tillkruss replied:
Sure. The condition needs to be tested, so we need a test where `$messages` is falsey and one where it's true. From what I can tell, you already have a test for when it's true, correct?

Eugene-Melbourne replied:
Actually `messages` neither true nor false it is a way to check if the property exists. The `messages` property is an object of `MessageBag` class. Usually I do checks like this using  `is_null($this->messages)`, but this time, I followed a style of the module.
The function `passes` - is the place where `messages` created and validation performed. It is also not great, because it comes with assumption that if messages are created than validation was performed. So you cannot just create messages without performing validation. And again, I go with the flow, performing validation as expected to create a `MessageBag` object for a `messages` property.

Eugene-Melbourne replied:
@tillkruss , You know what, I can make 2 test cases, one with `messages` created prior to calling addFailure  and one where no messages. I don't see a point doing it, I could not assert much with it, but maybe that is what is expected?

Eugene-Melbourne replied:
@tillkruss , Even better, I can call addFailure  2 times to asert that previous error messages were not lost. Would it count for what is expected?

Eugene-Melbourne committed:
removing return type declaration

removing return type declaration from test case
